File: python/day21.py
import numpy as np


with open('./inputs/21.txt') as f:
    instructions = f.readlines()

mappings = {}
start = '.#./..#/###'

def translate_to_np(s):
    return np.array([[c == '#' for c in l]
                     for l in s.split('/')])

for line in instructions:
    k, v = map(translate_to_np, line.strip().split(' => '))
    for a in (k, np.fliplr(k)):
        for r in range(4):
            mappings[np.rot90(a, r).tobytes()] = v


def find_next(grid):
    size = len(grid)
    by = 2 if size % 2 == 0 else 3
    new_size = size * (by+1) // by
    new_grid = np.empty((new_size, new_size), dtype=bool)
    squares = range(0, size, by)
    new_squares = range(0, new_size, by+1)

    for i, ni in zip(squares, new_squares):
        for j, nj in zip(squares, new_squares):
            square = grid[i:i+by, j:j+by]
            enhanced = mappings[square.tobytes()]
            new_grid[ni:ni+by+1, nj:nj+by+1] = enhanced
    return new_grid


grid = translate_to_np(start)
for _ in range(5):
    grid = find_next(grid)
print(grid.sum())



seen = {}
def find_new_9x9(grid):
    gtb = grid.tobytes()
    if gtb not in seen:
        for _ in range(3):
            grid = find_next(grid)
        seen[gtb] = grid
    return seen[gtb]


grid = find_next(grid) # 6th iteration -> 27x27 grid of independent 3x3's
for _ in range((18 - 6) // 3):
    size = len(grid)
    solution = np.empty((3*size, 3*size), dtype=bool)
    for i in range(0, size, 3):
        for j in range(0, size, 3):
            solution[3*i:3*(i+3), 3*j:3*(j+3)] = find_new_9x9(grid[i:i+3, j:j+3])
    grid = solution
print(grid.sum())

File: nim/day21.nim
import strutils, tables, sequtils

const instructions = readFile("./inputs/21.txt").splitLines
type Grid = seq[seq[bool]]


func flip(s: var Grid) = swap(s[0], s[^1])

func transpose(s: var Grid) =
  for i in 0 .. s.high:
    for j in i+1 .. s.high:
      swap(s[j][i], s[i][j])

iterator transform(s: var Grid): Grid =
  for _ in 1 .. 4:
    yield s; s.transpose
    yield s; s.flip

func toBool(s:string): seq[bool] =
  result = newSeq[bool](s.len)
  for i, c in s:
    result[i] = c == '#'


func parseInput(): Table[Grid, Grid] =
  result = initTable[Grid, Grid]()
  var
    divided: seq[string]
    source, dest: seq[seq[bool]]
  for line in instructions:
    divided = line.strip.split(" => ")
    source = divided[0].split('/').map(toBool)
    dest = divided[1].split('/').map(toBool)
    for src in source.transform:
      result[src] = dest


const
  startingGrid = ".#./..#/###".split("/").map(toBool)
  mappings = parseInput()


func getSquare(grid: Grid, i, j, by: int): Grid =
  result = newSeqWith(by, newSeq[bool](by))
  var row = 0
  for x in i ..< i+by:
    result[row] = grid[x][j ..< j+by]
    inc row


func findNext(grid: Grid): Grid =
  let
    size = grid.len
    by = if size mod 2 == 0: 2 else: 3
    newSize = size * (by+1) div by
    squareSize = size div by
  var square, enhanced: Grid
  result = newSeqWith(newSize, newSeq[bool](newSize))

  for i in 0 ..< squareSize:
    for j in 0 ..< squareSize:
      square = grid.getSquare(i*by, j*by, by)
      enhanced = mappings[square]
      for x, row in enhanced:
        for y, c in row:
          result[i*(by+1) + x][j*(by+1) + y] = c


func forward3steps(pattern: Grid): CountTable[Grid] =
  result = initCountTable[Grid]()
  var
    pattern = pattern
    square: Grid
  for _ in 1 .. 3:
    pattern = pattern.findNext
  for r in 0 .. 2:
    for c in 0 .. 2:
      square = pattern.getSquare(3*r, 3*c, 3)
      result.inc(square)


func pixelsOn(grid: Grid): int =
  for r in grid:
    for c in r:
      if c: inc result


func fastCount(grid: Grid, steps: int): int =
  var
    blockCounts = initCountTable[Grid]()
    mappingCounts = initTable[Grid, CountTable[Grid]]()
  blockCounts[grid] = 1

  for step in 1 .. steps:
    var tempBlockCounts = initCountTable[Grid]()
    for pattern, count in blockCounts.pairs:
      if pattern notin mappingCounts:
        mappingCounts[pattern] = forward3steps(pattern)
      for enhPattern, enhCount in mappingCounts[pattern].pairs:
        tempBlockCounts.inc(enhPattern, count*enhCount)
    blockCounts = tempBlockCounts

  for pattern, count in blockCounts.pairs:
    result += pattern.pixelsOn * count


var first = startingGrid
for _ in 1 .. 5:
  first = first.findNext
echo first.pixelsOn

echo startingGrid.fastCount(18 div 3)
